pkg/client/unversioned/discovery_client.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unversioned
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+// DiscoveryClient implements the discovery functions.
+type DiscoveryClient struct {
+	http.Client
+	config Config
+}
+
+// ServerAPIVersions returns the GroupVersions supported by the API server.
+// It creates a RESTClient based on the passed in config, but it doesn't rely
+// on the Version, Codec, and Prefix of the config, because it uses AbsPath and
+// takes the raw response.
+func (d *DiscoveryClient) ServerAPIVersions() (groupVersions []string, err error) {
+	client := d.Client
+
+	if d.config.Version != "" || d.config.Prefix != "" {
+		return nil, fmt.Errorf("The config of a DiscoveryClient should have empty version and prefix")
+	}
+	baseURL, err := defaultServerUrlFor(d.config)
+	if err != nil {
+		return nil, err
+	}
+	// Get the groupVersions exposed at /api
+	baseURL.Path = "/api"
+	resp, err := client.Get(baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	var v unversioned.APIVersions
+	defer resp.Body.Close()
+	err = json.NewDecoder(resp.Body).Decode(&v)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+
+	groupVersions = append(groupVersions, v.Versions...)
+	// Get the groupVersions exposed at /apis
+	baseURL.Path = "/apis"
+	resp2, err := client.Get(baseURL.String())
+	if err != nil {
+		return nil, err
+	}
+	var apiGroupList unversioned.APIGroupList
+	defer resp2.Body.Close()
+	err = json.NewDecoder(resp2.Body).Decode(&apiGroupList)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error: %v", err)
+	}
+	groupVersions = append(groupVersions, extractGroupVersions(&apiGroupList)...)
+
+	return groupVersions, nil
+}
+
+func (d *DiscoveryClient) SupportedResourcesForGroupVersion(groupVersion string) (*api.APIResourceList, error) {
```

Comment:
This is copied from Brendan's #15427. I will change the names after that one get merged.